# <font color='red'><b>Bootstrap Implementation in Random Forest</b> </font>

## Importing packages

In [ ]:
import numpy as np # importing numpy for numerical computation
from sklearn.datasets import load_boston # here we are using sklearn's boston dataset
from sklearn.metrics import mean_squared_error # importing mean_squared_error metric

In [ ]:
# load the data
boston = load_boston()
x=boston.data #independent variables
y=boston.target #target variable

In [ ]:
x.shape

(506, 13)

In [ ]:
# display part of the data
x[:5]

array([[6.3200e-03, 1.8000e+01, 2.3100e+00, 0.0000e+00, 5.3800e-01,
        6.5750e+00, 6.5200e+01, 4.0900e+00, 1.0000e+00, 2.9600e+02,
        1.5300e+01, 3.9690e+02, 4.9800e+00],
       [2.7310e-02, 0.0000e+00, 7.0700e+00, 0.0000e+00, 4.6900e-01,
        6.4210e+00, 7.8900e+01, 4.9671e+00, 2.0000e+00, 2.4200e+02,
        1.7800e+01, 3.9690e+02, 9.1400e+00],
       [2.7290e-02, 0.0000e+00, 7.0700e+00, 0.0000e+00, 4.6900e-01,
        7.1850e+00, 6.1100e+01, 4.9671e+00, 2.0000e+00, 2.4200e+02,
        1.7800e+01, 3.9283e+02, 4.0300e+00],
       [3.2370e-02, 0.0000e+00, 2.1800e+00, 0.0000e+00, 4.5800e-01,
        6.9980e+00, 4.5800e+01, 6.0622e+00, 3.0000e+00, 2.2200e+02,
        1.8700e+01, 3.9463e+02, 2.9400e+00],
       [6.9050e-02, 0.0000e+00, 2.1800e+00, 0.0000e+00, 4.5800e-01,
        7.1470e+00, 5.4200e+01, 6.0622e+00, 3.0000e+00, 2.2200e+02,
        1.8700e+01, 3.9690e+02, 5.3300e+00]])

## Part 1

### <font color='red'> <b>Step - 1</b></font>

*  <b>Procedure for Creating samples</b><br>
    <b> Randomly create 30 samples from the whole boston data points</b>
    *  Creating each sample: Consider any random 303(60% of 506) data points from whole data set and then replicate any 203 points from the sampled points
    
     For better understanding of this procedure lets check this examples, assume we have 10 data points [1,2,3,4,5,6,7,8,9,10], first we take 6 data points randomly , consider we have selected [4, 5, 7, 8, 9, 3] now we will replicate 4 points from [4, 5, 7, 8, 9, 3], consder they are [5, 8, 3,7] so our final sample will be [4, 5, 7, 8, 9, 3, 5, 8, 3,7]
* <b> Create 30 samples </b>
    *  Note that as a part of the Bagging when we are taking the random samples <b>we have to make sure each of the sample will have different set of columns</b><br>
Ex: Assume we have 10 columns[1 ,2 ,3 ,4 ,5 ,6 ,7 ,8 ,9 ,10] for the first sample we will select [3, 4, 5, 9, 1, 2] and for the second sample  [7, 9, 1, 4, 5, 6, 2] and so on...
Each sample will have atleast 3 feautres/columns/attributes

* <font color='red'><b> Note - While selecting the first random 60% datapoints from the whole data, we are to make sure that the selected datapoints are all exclusive, repetition is not allowed. </b></font>

### <font color='red'><b>Step - 2 </b></font>

### Building High Variance Models on each of the sample and finding train MSE value

*  Build a regression trees on each of 30 samples.
*  Computed the predicted values of each data point(506 data points) in your corpus.
*  Predicted house price of $i^{th}$ data point $y^{i}_{pred} =  \frac{1}{30}\sum_{k=1}^{30}(\text{predicted value of } x^{i} \text{ with } k^{th} \text{ model})$
*  Now calculate the $MSE =  \frac{1}{506}\sum_{i=1}^{506}(y^{i} - y^{i}_{pred})^{2}$

###<font color='red'> <b>Step - 3 </b></font>

### Calculating the OOB score

*  Predicted house price of $i^{th}$ data point $y^{i}_{pred} =  \frac{1}{k}\sum_{\text{k= model which was buit on samples not included } x^{i}}(\text{predicted value of } x^{i} \text{ with } k^{th} \text{ model})$.
*  Now calculate the $OOB Score =  \frac{1}{506}\sum_{i=1}^{506}(y^{i} - y^{i}_{pred})^{2}$.

# Part 2

*  Computing CI of OOB Score and Train MSE
  *   We will repeat Part 1 for 35 times, and for each iteration store the Train MSE and OOB score.
<li> After this we will have 35 Train MSE values and 35 OOB scores </li>
<li> using these 35 values we will find the confidence intravels of MSE and OOB Score </li>
<li> We will print the CI of MSE and CI of OOB Score </li>
</ol>

# Part 3

*  Given a single query point predict the price of house.

Consider xq= [0.18,20.0,5.00,0.0,0.421,5.60,72.2,7.95,7.0,30.0,19.1,372.13,18.60] 
Predict the house price for this point as mentioned in the step 2 of Task 1.

## <font color='red'><b>A few key points</b></font>

* Remember that the datapoints used for calculating MSE score contain some datapoints that were initially used while training the base learners (the 60% sampling). This makes these datapoints partially seen (i.e. the datapoints used for calculating the MSE score are a mixture of seen and unseen data).
Whereas, the datapoints used for calculating OOB score have only the unseen data. This makes these datapoints completely unseen and therefore appropriate for testing the model's performance on unseen data.

* Given the information above, if your logic is correct, the calculated MSE score should be less than the OOB score.

* The MSE score must lie between 0 and 10.
* The OOB score must lie between 10 and 35.

* The difference between the left nad right confidence-interval values must not be more than 10. Make sure this is true for both MSE and OOB confidence-interval values.

# Implementing Part - 1

* Creating samples

<font color='Orange'><b>Algorithm</b></font>

![alt text](https://i.imgur.com/OfcFrUP.jpg/)

In [ ]:
def generating_samples(input_data, target_data):

    '''In this function, we will write code for generating 30 samples '''
    
    # We will follow above pseudo code for generating samples 
    
    selecting_rows, selecting_columns = [], []   # Initialising the lists to store the column and row samples
    
    # First we need to select the random 60% data points(or indices here) "without replacement"
    selecting_rows = np.random.choice(np.arange(0, len(input_data)),size=303, replace=False)
    
    # Selecting rows with replacement from the above list
    replacing_rows = np.random.choice(np.arange(0,len(selecting_rows)),size=203, replace=True)
    
    # Now selecting "unique" column indices, i.e. SRSWOR
    selecting_columns = np.random.choice(np.arange(0, 13),size=np.random.randint(low=3, high=14), replace=False)
    
    # selecting the data from our dataset based on the obtained indices
    sample_data = input_data[selecting_rows[:,None],selecting_columns]
    
    target_of_sample_data = target_data[selecting_rows]
    
    replicated_sample_data = sample_data[replacing_rows]
    
    target_of_replicated_sample_data = target_of_sample_data[replacing_rows]
    
    # Stack all the data and return them
    final_sample_data = np.vstack((sample_data,replicated_sample_data))
    final_target_data = np.vstack((target_of_sample_data.reshape(-1,1),target_of_replicated_sample_data.reshape(-1,1)))
    
    return list(final_sample_data), list(final_target_data), list(selecting_rows), list(selecting_columns)

    # return sampled_input_data , sampled_target_data,selected_rows,selected_columns
    #note please return as lists

*  Create 30 samples

![alt text](https://i.imgur.com/p8eZaWL.jpg)

In [ ]:
# Use generating_samples function to create 30 samples 
# store these created samples in a list
list_input_data =[]
list_output_data =[]
list_selected_row= []
list_selected_columns=[]

for i in range(30):
    a,b,c,d = generating_samples(x, y)
    list_input_data.append(a)
    list_output_data.append(b)
    list_selected_row.append(c)
    list_selected_columns.append(d)
    

<font color='orange'><b>Flowchart for building tree</b></font>

![alt text](https://i.imgur.com/pcXfSmp.png)

In [ ]:
from sklearn.tree import DecisionTreeRegressor

# we will use a dictionary to store the all the models
models = []
for i in range(30):
    
    # We will fit our decision tree model with the data generated above
    # Declaring our decision tree regressor
    clf = DecisionTreeRegressor(max_depth=None)
    clf.fit(list_input_data[i],list_output_data[i])
    
    # Storing the models in a list for further use
    models.append(clf)

<font color='orange'><b>Flowchart for calculating MSE </b></font>

![alt text](https://i.imgur.com/sPEE618.png)

After getting predicted_y for each data point, we can use sklearns mean_squared_error to calculate the MSE between predicted_y and actual_y.

In [ ]:
# Creating an empty list to store the predicted values of y
predicted_y = []

for i in range(len(x)):
    
    # creating a temporary list to store the outputs of 30 models which will then be aggregated using median
    pred_y = []
    # Iterating over all the saved models to get prediction for a paricular datapoint
    for j in range(30):
        # Selecting the columns for the given model
        datapoint = x[i,list_selected_columns[j]]
        
        pred_y.append(models[j].predict(datapoint.reshape(1,-1)))
    
    # Using the median to aggregate all the outputs
    predicted_y.append(np.median(pred_y))

# calculating MSE score
mse_score = mean_squared_error(y,predicted_y)
print(mse_score)

0.15115013781193626


<font color='orange'><b>Flowchart for calculating OOB score</b></font>

![alt text](https://i.imgur.com/95S5Mtm.png)

Now calculate the $OOB Score =  \frac{1}{506}\sum_{i=1}^{506}(y^{i} - y^{i}_{pred})^{2}$.

* Code for calculating OOB score

In [ ]:
# Calculating OOB score

# Creating a list to store the predicted values of oob data points
predicted_oob_y = []

for i in range(len(x)):
    
    # creating a temporary list to store predicted values from all the models for a particular datapoint
    pred_oob_y = []
    
    for j in range(30):
        
        # Checking if the index "i" is present in the list of selected rows corresponding to the model j
        if i not in list_selected_row[j]:
            # Extract the columns required for the jth model
            datapoint = x[i,list_selected_columns[j]]
            
            # storing the predicted y for each model
            pred_oob_y.append(models[j].predict(datapoint.reshape(1,-1)))
    
    # appending the median value of the predicted base y's to get the final predicted value for ith datapoint
    predicted_oob_y.append(np.median(pred_oob_y))

# calculating the OOB score as per the given formula
oob_score = mean_squared_error(y, predicted_oob_y)
print(oob_score)

13.36445234045715


# Part 2

In [ ]:
# find Confidence Interval of MSE and OOB_score

from tqdm import tqdm
# Appending all the subsections of task1 in one single cell
mse, oob = [], []

for i in tqdm(range(35)):
    
    # ********************** Generating Bootstrap Samples *************************************************
    list_input_data_1 = []
    list_output_data_1 = []
    list_selected_row_1 = []
    list_selected_columns_1 = []
    
    for _ in range(30):
        a,b,c,d = generating_samples(x, y)
        list_input_data_1.append(a)
        list_output_data_1.append(b)
        list_selected_row_1.append(c)
        list_selected_columns_1.append(d)
        
    
    # ********************** Base Model creation *************************************************
    
    # we will use a dictionary to store the all the models
    base_models = []
    for i in range(30):

        # We will fit our decision tree model with the data generated above
        # Declaring our decision tree regressor
        clf = DecisionTreeRegressor(max_depth=None)
        clf.fit(list_input_data_1[i],list_output_data_1[i])

        # Storing the models in a list for further use
        base_models.append(clf)
    
    # ******************** MSE Calculation ********************************************************
    
    # Creating an empty list to store the predicted values of y
    predicted_y = []

    for i in range(len(x)):

        # creating a temporary list to store the outputs of 30 models which will then be aggregated using median
        pred_y = []
        # Iterating over all the saved models to get prediction for a paricular datapoint
        for j in range(30):
            # Selecting the columns for the given model
            datapoint = x[i,list_selected_columns_1[j]]

            pred_y.append(base_models[j].predict(datapoint.reshape(1,-1)))

        # Using the median to aggregate all the outputs
        predicted_y.append(np.median(pred_y))

    # calculating MSE score
    mse_score = mean_squared_error(y,predicted_y)
    mse.append(mse_score)


    # ***************************** OOB calcualtion ******************************************
    
    # Creating a list to store the predicted values of oob data points
    predicted_oob_y = []

    for i in range(len(x)):

        # creating a temporary list to store predicted values from all the models for a particular datapoint
        pred_oob_y = []

        for j in range(30):

            # Checking if the index "i" is present in the list of selected rows corresponding to the model j
            if i not in list_selected_row_1[j]:
                # Extract the columns required for the jth model
                datapoint = x[i,list_selected_columns_1[j]]

                # storing the predicted y for each model
                pred_oob_y.append(base_models[j].predict(datapoint.reshape(1,-1)))

        # appending the median value of the predicted base y's to get the final predicted value for ith datapoint
        predicted_oob_y.append(np.median(pred_oob_y))

    # calculating the OOB score as per the given formula
    oob_score = mean_squared_error(y, predicted_oob_y)
    oob.append(oob_score)

    
# using the samples to calculate CI for MSE and OOB score with the below formula
# ci_of_x = [mean(x)-1.96*SE(x),mean(x)+1.96*SE(x)]

ci_of_mse = [np.mean(mse)-1.96*(np.std(mse)/np.sqrt(len(mse))),np.mean(mse)+1.96*(np.std(mse)/np.sqrt(len(mse)))]

ci_of_oob = [np.mean(oob)-1.96*(np.std(oob)/np.sqrt(len(oob))),np.mean(oob)+1.96*(np.std(oob)/np.sqrt(len(oob)))]

print(f"CI of MSE is: ({ci_of_mse[0]}, {ci_of_mse[1]})")
print(f"CI of OOB Score is: ({ci_of_oob[0]}, {ci_of_oob[1]})")

100%|██████████████████████████████████████████████████████████████████████████████████| 35/35 [02:05<00:00,  3.59s/it]

CI of MSE is: (0.06273894679611279, 0.10581349757281876)
CI of OOB Score is: (12.840101088316112, 13.844243657079774)


# <font color='blue'><b>Task 3</b></font>

<font color='orange'><b>Flowchart for Task 3</b></font>

<b>Hint: </b> We created 30 models by using 30 samples in TASK-1. Here, we need send query point "xq"  to 30 models and perform the regression on the output generated by 30 models.

![alt text](https://i.imgur.com/Y5cNhQk.png)

*  <font color='blue'><b> Write code for TASK 3 </b></font>

In [ ]:
# predict the yq-value for the given xq

xq = np.array([0.18,20.0,5.00,0.0,0.421,5.60,72.2,7.95,7.0,30.0,19.1,372.13,18.60])
y_vals = []

for i in range(30):
    datapoint = xq[list_selected_columns[i]]
    y_vals.append(models[i].predict(datapoint.reshape(1,-1)))
print(f'Predicted Y = {np.median(y_vals)}')

Predicted Y = 19.4


<br><br><br>

<font color='red'><b>Write observations for task 1, task 2, task 3 indetail</b></font>

### Task 1
<br></br>
#### Step 1 - Generating Bootstrap Samples
* For Step 1 we were focussed on re-creating the concept of bootstrap samples from the original dataset. The size of the bootstrap samples should be same as that of the original dataset. 
- For this first sample we selected 60% of the datapoints from the original dataset ***without replacement***. Let's call this sample1. 
* Now using sample1 from the above step we again perform sampling ****with replacement**** for the remaining 40% of the points of our new dataset.
* We stack this two samples together to form a new dataset which is same in size as our original dataset.
* Apart form row sampling we are also performing column sampling(w/o replacement) on our available dataset( one key aspect in RF).
* Now we repeat this above process 30 times to obtain our 30 Bootsrap samples.

<br></br> 
#### Step 2: Building High variance models and Calculating MSE
* In this step we are going to build 30 high variance base models, DT Regressors(max_depth=None) one for each of the samples generated in the above step. We are also storing the models so that we can use them further.
* After the models are fit we are going to use it to calculate the train MSE.
+ To calculate train MSE we pass all the points in our train data through all the models. Lets take the example of one datapoint passing through this RF model.
    * For a single point say x', first we are passing it through all the 30 DT models we have.
    * Where for each model different columns of x' may be used. We apply the columns selected in column sampling as required for whichever model and pass that point to the corresponding model.
    * Each model gives us a predicted y i.e. we have a list of 30 predicted y's.
    * Now we aggregate all these 30 predicted y's using either mean or median.
    * This gives us the y' corresponding to x'.
* So when we pass out dataset through the model, we get a list of predicted values of x, the list of predicted y-vals.
* Using this list of predicted y-vals, and the original/target y-vals we are going to calculate MSE, which is the average of sum of deviations square.

<br></br> 
#### Step 3: Calculating OOB Score
* For calculating OOB Score we are going to follow the following steps:
    * for a particular datapoint xi, check in which base model this was **not used in the training stage**. In whichever model this datapoint was not used in the training stage, use that model to predict its y-value( after applying the column sampling)
    * Now aggregate all the y-values generated(which will be less than 30) to get the final yi.
* We repeat this step for all the datapoints in our dataset to get a list of predicted OOB y-vals.
* Finally we use the original y-vals and predicted OOB y-vals to obtain our OOB Score as per the formula.


### Task 2 : Calculate CI of MSE and OOB Score
<br></br>
* We are going to repeat Task1 35 times to get a list of 35 MSE and OOB score.
* Assuming this as a sample of MSE and OOB Score we are going to use it to calculate our 95% Confidence Interval.
* For calculation of CI we need to calculate the mean and Standard Errow of MSE and OOB Scores.
    * For lower limit of CI of MSE: $ mean(MSE) - 1.96 * SE(MSE)/  sq.root(n) $ <br></br>
    * For upper limit of CI of MSE: $ mean(MSE) + 1.96 * SE(MSE)/  sq.root(n) $ <br></br>
where n is the number of datapoints in the sample

* We can repeat the same for OOB score values, which will give us the lower and the upper limit of CI OOB Score

### Task 3 : Predict yq given xq.

* Given a query point we have to pass it through all the base models and get the predicted value from each model and the aggregate them using median to get our final predicted y value.